### PR TITLE
Implement subtasks

### PR DESCRIPTION
**Pull request by raimund-schluessler**
Things to do:

- [x] enable moving subtasks to different parent task
- [x] always load completed tasks with uncompleted root or parent task
- [x] completing a parent tasks has to complete all subtasks
- [x] uncompleting a subtask has to uncomplete the parent task
- [x] deleting a task must delete all related subtasks
- [x] adding a subtask to a completed parent must uncomplete parent
- [x] consider how to handle subtasks with available root task but missing parent task (due to shared/private calendars) --> They will be shown as root task, because the link between these tasks cannot be resolved due to the missing (private) parent task.
- [x] add subtasks by input field below task after clicking a Plus icon
- [x] ```hasNoParent``` must account for tasks whose parents are not loaded (due to shared / private calendars)
- [x] style subtasks list
- [x] add an icon to show existence of subtasks
- [x] enable hiding of subtasks
- [x] only count root tasks
- [x] fixing searching for subtasks
- [x] load rootTask when querying for child task in ```TaskController::getTask()```
- [x] refactor Drag and Drop of tasks

### CODE REVIEW

**Comment by raimund-schluessler** [comment]
@jancborchardt @Henni 
I would like to have some feedback about how to handle these points:

1) How to handle the completing of a parent task?
I think all subtasks should be completed to, as a parent tasks cannot be completed without completing all children. Also for clients not knowing subtasks this cleans the tasks list, otherwise children would still be shown in the todo list of other clients.

2) How to handle the uncompleting of a child task?
I think uncompleting a child task should also uncomplete the parent task for the same reasons as in 1).

3) How to handle the deleting of a parent task?
We could either delete all its children too or make the children the new parent tasks (or we could ask the user what he wants). I am not sure about this one, though.

Opinions?

**Comment by Henni** [comment]
Looks great so far.

Regarding your questions:
**1) How to handle the completing of a parent task?**
I think child tasks should be completed if the parent task is completed. But parent tasks shouldn't be completed if all subtasks are completed, because maybe you want to add another subtask later on.
**2) How to handle the uncompleting of a child task?**
Uncompleting a child task should in my opinion always uncomplete the parent task.
**3) How to handle the deleting of a parent task?**
I would delete all the child tasks. I think this is intuitive and if the user wants to convert the child tasks to new root tasks he is able to do this manually.

I think I found an usability problem. I think we have to find a different way of adding subtasks.
Currently a new task is automatically added to the active task. Root tasks are only added if there is currently no active task.
I think this isn't intuitive and we should replace it maybe with an input right next to or under the active task.

Do we want to limit the number of possible hierarchy levels? I believe we don't need subtasks of subtasks.

**Comment by raimund-schluessler** [comment]
Thanks for the comments.

1) Yes, completing all subtasks will not complete the parent.
2) Alright :)
3) I will implement it this way, then we can have a look again.

Hm, we could show an input field below the active task when the user clicks e.g. on a Plus icon within the task body. Always showing it might be ugly.

There is no reason to limit the hierarchy level. Subtasks are just normal tasks and can have Subtasks itself, thats how the CalDAV specification handles it. And if other clients create a task with a parent which is a subtask we have a problem showing this, when we limit ourself.

**Comment by sahne123** [comment]
Things going more complicated with private and confidential tasks in a shared calendar. If there are subtasks private or confidential you shouldn't be allow to complete these tasks by completing the parent. 

By completing the parent task we might complete all subtasks with permissions and check afterwards if there are any uncompleted left before we can complete the parent task.

But how to handle the other way? Confidential or even private parent task with public subtasks. Should we show these at all?

This cases may be theoretical but i think we should follow permissions exactly. So we can't be sure that all subtasks are complete if the parent is complete.

**Comment by raimund-schluessler** [comment]
We will show all tasks which are accessible to a user. If a task is private it must not be shown. If it has public subtasks, then these subtasks will appear, but as normal tasks without a parent task. For the backend all tasks are equal, only the GUI decides how to show them. So when editing a task (completing, etc.) the backend decides if this is allowed and handles it accordingly. We will also try to complete all children, but if they are read-only or private we can't do so.

We might have to fix the consideration of the permissions, but it won't get any more complicated by subtasks.

**Comment by raimund-schluessler** [comment]
Adding subtasks via the input fields below the parent task is not functional yet, but the visual impression might be correct.

**Comment by raimund-schluessler** [comment]
Adding subtasks by this input field should work now.

**Comment by sahne123** [comment]
@raimund-schluessler agree. We just have to keep in mind (for displaying) there could be incomplete subtask within a completed parent task. will have a look at your work asap.

**Comment by raimund-schluessler** [comment]
Currently its handled like this:

If the root task is completed, the root task and its subtasks are shown in the completed section whether or not there are uncompleted subtasks. This might be changed, although such cases should only occur if you don't have write access to one of the subtasks.

If one task is actually a subtask of a parent which is private, then the parent should not be loaded obviously and the subtask is shown as a root task. We should consider what to do if you have access to the root task and e.g. second level subtasks, but not to the first level subtasks.

Most of this is a question about how we load tasks in the ```tasksservice.php``` file here:
https://github.com/owncloud/tasks/blob/master/service/tasksservice.php#L21

**Comment by Henni** [comment]
Good work so far!

some thoughts:
* I have a high dpi display and because of that i set the default zoom inside my browser to 150%. This leads to graphical glitch where the subtasks overlap. Is there a way to fix this with some css magic? @jancborchardt do you have an idea?
overlapping_subtasks
* I think an offset of 35px would look better. This way it is also aligned with the task seperator which is transparent BTW (should we simply remove the seperator?)
image
* The current way of adding subtasks looks good, but won't work with touch inputs. You would have to guess where to click because the button only shows up on hover.

**Comment by raimund-schluessler** [comment]
- I will have a look, but I don't know if I can reproduce this, since my display is not high dpi.

- I will adjust the offset, sounds reasonable to align it there. The separator can be removed.

- Is there a way to always show it on touch displays by media queries of something like that? I would like to keep the hover to not clutter up the task-body.

**Comment by Henni** [comment]
Regarding the third point:
Media queries are currently only supported by chrome and opera: http://caniuse.com/#feat=css-media-interaction
Firefox has something similar: https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Media_queries#-moz-touch-enabled

We could use modernizr to check for touch support but this doesn't imply that the user actually uses touch. http://modernizr.com/docs/#features-misc

Maybe the best solution would be to also list the subtasks in the details view so that a user on a mobile device has an alternative way of adding subtasks.

**Comment by raimund-schluessler** [comment]
Thanks for the links.

However, adding further library just to check for touch, which isnt even perfect, is a bit exaggerated.

Showing the subtasks in the detail view is also not a good solution I think.

I will just remove the hover effect, and if we think about a better solution later on, we can still implement it.

**Comment by Henni** [comment]
sounds good to me

**Comment by jancborchardt** [comment]
Do we really want so many sublevels of subtasks? One or two levels seems to be enough. (I know more is possible, but how much makes sense? So far there weren’t any subtasks at all so we shouldn’t go all out on them now.)

**Comment by Henni** [comment]
@jancborchardt we talked about this a few comments ago ;)
I think @raimund-schluessler has a valid point:
> There is no reason to limit the hierarchy level. Subtasks are just normal tasks and can have Subtasks itself, thats how the CalDAV specification handles it. And if other clients create a task with a parent which is a subtask we have a problem showing this, when we limit ourself.

Maybe it makes sense to limit subtask creation, but we shouldn't limit how many sublevels are displayed.

**Comment by jancborchardt** [comment]
Well, we already basically have one level in the form of task lists. I’d say for now we should limit creation to one level.

**Comment by jancborchardt** [comment]
We could display stuff otherwise if it’s created by other clients, yeah.

**Comment by raimund-schluessler** [comment]
I think we need to display it correctly with subtasks of subtasks. So it is only the question whether or not we show the input field.

**Comment by raimund-schluessler** [comment]
This css glitch only happens for certain zoom-levels for me. Seems like rendering problem within the browser.

**Comment by Henni** [comment]
@raimund-schluessler are you aware that it is currently not possible to add root tasks?
image

**Comment by raimund-schluessler** [comment]
Thanks for the hint, didn't notice that.

**Comment by Henni** [comment]
@raimund-schluessler will this be in the next release? We should probably release 0.7 on 7th of July (the release date of owncloud 8.1)

**Comment by raimund-schluessler** [comment]
I would like to see it in the next release. However, there are still some things to do and I don't know if I can finish it in time.

We can either release without this feature, or wait another few days (I would like to test it on my production server anyway, but I will upgrade only after 8.1. is released)

**Comment by raimund-schluessler** [comment]
@Henni Please have a look at the latest status of this PR here. I implemented changing the tasks parent. You can drag and drop tasks just like before, changing the completed state or its calendarID. Hovering over a task for more than 1 s and dropping it then, makes the hovered task the new parent of the dragged task.

Would be good, if you could play around and test it a bit. Maybe we can make it for next release.

**Comment by raimund-schluessler** [comment]
PS: I don't understand scrutinizer. Why is the code quality getting worse by 0.45 points??

**Comment by Henni** [comment]
> Hovering over a task for more than 1 s and dropping it then, makes the hovered task the new parent of the dragged task.
Why do you wait a second?

Sometimes I get this error, but I don't know how to reproduce it. Maybe this happens with tasks which were created in a previous commit or tasks from different lists.
image

Should we restrict the number of sublevels the user is allowed to create?

**Comment by raimund-schluessler** [comment]
> Why do you wait a second?

I wait a second so the user can decide whether he wants to make it a subtask (after one second), or he just wants to change the list, complete/uncomplete or star a task (the behavior as it was before subtasks were implemented) or make it a root task.

> Sometimes I get this error, but I don't know how to reproduce it. Maybe this happens with tasks which were created in a previous commit or tasks from different lists.

Could you post the line at which this error occurs? Maybe I can track it down.

> Should we restrict the number of sublevels the user is allowed to create?

I am still a bit reluctant to this, since I am not convinced that the users have problems with deeper nested subtasks. I find it intuitiv to have subtasks of subtasks and not confusing. Also it is quite some work to limit it (I think).

**Comment by Henni** [comment]
How do you change completeness or star/instar by drag and drop. 
In which use case will the user drop the task without wanting it to be a subtask? Or do you wait the second so that the user is able to abort the action?


The error occurs on line 557 of app.js

**Comment by raimund-schluessler** [comment]
Sorry, you are right, starring/unstarring is not possible by dropping onto a task. This is only for dropping onto the starred list at the left.

However, you can complete a task when you have selected a calendar at the left. Then you see the uncomplete tasks at the top and the completed ones below. You can drop tasks onto these regions and they get completed or uncompleted (just try it out). Another reason for waiting is that the user probably wants to make a root tasks out of a subtask or he just wants to switch calendars (e.g. in the 'All' overview). This is possibly by dropping it before one second has passed.

Are you sure, it is line 557? At this line there is no ```related``` property set...
https://github.com/owncloud/tasks/blob/Subtasks/js/public/app.js#L557

**Comment by Henni** [comment]
Oh this was the line 557 in the app.js of the appframework vendor script.
You should be able to reproduce it by moving tasks between different lists in the 'all' view.

I'm not sure whether the one second delay is intuitive.
* When you are in the view for a single list nothing happens if you drop the task during the delay.
* Completing a task and switching the calendar is possible by dropping on the navigation entry.
* How do I convert a subtask to a root task?

**Comment by raimund-schluessler** [comment]
I will have a look, thanks!

You can convert a subtask to a root task by dragging it over an other task and dropping it before the 1 second delay.

**Comment by Henni** [comment]
This doesn't work for me!

**Comment by jancborchardt** [comment]
Btw @Henni, just saw your indentation alignment comment above – great call on aligning the checkbox of subtasks with the text of the parent task! :)

**Comment by raimund-schluessler** [comment]
@Henni I think you are right, the drag and drop implementation needs to be revised. The user should instantly see what will happen when he drops the task. So it should be temporally shown within the task list while it is dragged and before it is dropped. Realizing it this way could also be used as basement for manually sorting the task list.

Since this means a lot of additional work for the subtasks feature, we should delay it for v0.8 of the app. I think with category and priority support v0.7 has two nice features already ;)

**Comment by Henni** [comment]
@raimund-schluessler sounds like a good decision.

So let's release a beta tomorrow with owncloud 8.1 and the final release later this week. Do you agree?

**Comment by raimund-schluessler** [comment]
Fine with me. :+1:

**Comment by Henni** [comment]
Great to see this code cleanup! (especially the date duplication)
Is there a specific reason why you added this to the subtasks branch instead of master?

**Comment by raimund-schluessler** [comment]
Well, I actually should have added this to the master branch, or at least a separate new branch.

I was working on the implementation of only loading uncompleted root tasks but all of their children (whether completed or not) to solve one of the remaining issues. But this created some very unclear functions. Hence, I started to refractor ```TasksService::getAll()``` and it became a bigger issue. And then I thought if I refactor this, I can just as well also refactor the rest...

Probably not my best decision, but I guess it is to much work to split this into separate branches now (and probably not necessary once Subtasks is merged).

**Comment by Henni** [comment]
ok :+1: 
I'm not completely happy with the Helper Class yet. Do you have an idea how to clean/split it up?

**Comment by raimund-schluessler** [comment]
Well, the functions related to the reminder (```Helper::parseReminder()``` and ```TasksService::setReminderDate()```) and ```Helper::updateVCalendarFromRequest()``` need to be simplified / split. I will have a look at this too, when the rest is pretty enough ;)

**Comment by raimund-schluessler** [comment]
Well, the current implementation works... But on my development server it takes very long to load (3~4s), and it takes literally forever on my productive server with many calendar entries (I stopped the request after 2 min).
There got to be a faster way to get the child tasks then looping through all tasks again and again (current implementation of ```Helper::getChildTasks()```).

**Comment by raimund-schluessler** [comment]
Does anybody know how I can access the database mapper https://github.com/owncloud/tasks/blob/Subtasks/db/tasksmapper.php in the searchcontroller https://github.com/owncloud/tasks/blob/Subtasks/controller/searchcontroller.php#L42 ? Injecting it into the constructor like for the tasksservice https://github.com/owncloud/tasks/blob/Subtasks/service/tasksservice.php in the application controller https://github.com/owncloud/tasks/blob/Subtasks/appinfo/application.php#L95 does not seem to be possible? Do I miss something? Is this possible at all?

Maybe @butonic or @Raydiation know? Thanks for the help, all tips are welcome ;)

**Comment by Raydiation** [comment]
A few things: Do not directly access the database layer in your controller. 

Why? Because your database layer does not contain validation and the validation needs to go into the controller then. 
Why is this bad? Because you want slim controllers. 
Why do you want slim controllers? Because a controller is essentially the glue between HTTP and your app and is harder to reuse and test because you need to deal with HTTP stuff.

-> Create a service class instead that calls your database layer methods

Leaving that aside, there are several things wrong with the SearchController code: 
* It inherits from the OCP\Search\Provider class which is not a controller. This code will actually not work. Instead create a controller and a service which inherits from the provider class, then inject that 
* \OC::$server->getUserSession()->getUser()->getUID() should be injected. There is a shortcut called $userId which can just be set in the constructor.

IMHO create a search service class and inject that into the tasks controller since searching is usually an index method, e.g. **/authors?search=name:test,something:else**, therefore it's actually bad practice to create a separate controller just for searching.

Lastly, please go through the tutorial because I think you did not fully grasp how everything works together ;) https://doc.owncloud.org/server/8.1/developer_manual/app/tutorial.html

**Comment by Raydiation** [comment]
BTW: you can also just ping me on irc.freenode.net #owncloud-dev for further questions ;)

**Comment by raimund-schluessler** [comment]
Thank you very much for the insight :) I already had a look at the tutorial, but it seems like I really did not fully understood how everything works. Especially the global search as there is little to none explanations how to implement the global search for an app (or at least I didn't find them) ;)

One question still remains:
When I create a search service and inject it into the TasksController, does this mean, I have to register the TasksController as a SearchProvider here https://github.com/owncloud/tasks/blob/Subtasks/appinfo/application.php#L137? How can I do this? The line
```php
\OC::$server->getSearch()->registerProvider('OCA\Tasks\Controller\SearchController', array('apps' => array('tasks')));
```
doesn't seem to be right, as I can't inject anything?

Also thanks for the IRC invitation, but I still have to set this up first ;)

**Comment by Raydiation** [comment]
The preferred way for injecting (since 8.0) is to just use the interface/class name and do not care where it comes from (also known as Hollywood Principle: Don't call us, we call you)

Classes can be injected by simply writing down the type hint, e.g.:

```php
public function __construct(YourClass $class)
```

This will instantiate YourClass once for the whole request and pass it in. This works if you use the convention over configuration approach, otherwise you'll have to add for the controllers it in the container setup, e.g.:

```php
$container->registerService('TheController', function ($c) {
    return new TheController($c->query('OCA\YourApp\YourClass'));
})
```
This is redundant as mentioned above.

Now for interfaces: they can't be instantiated so you have to tell the container which class should be used for an interface. If you take a look at the server, it says it returns an OCP\ISearch: https://github.com/owncloud/core/blob/master/lib/private/server.php#L719 . So we have to wire up a class for that: https://github.com/owncloud/core/blob/c19f0f38f3a097bfa111c91c00e6ae95616e8cc6/lib/private/appframework/dependencyinjection/dicontainer.php#L191

So you see this has already been done for you. If you want to register your own interface, use something like this in your application.php:

```php
$container->registerService('OCA\YourApp\IYourInterface', function ($c) {
    return $c->query('OCA\YourApp\YourClassWhichImplementsYourInterface')';
})
```
Note that this could be made easier by adding a new method to the SimpleContainer class (feel free to do a pull request xD), e.g.:

```php
$container->registerAlias('OCA\YourApp\IYourInterface', 'OCA\YourApp\YourClassWhichImplementsYourInterface');
```

Ok, so back to your initial question: The question is basically where the search provider should be registered? Globally? Only for specific routes?

I'd probably add a method to the Application class since its probably used globally:

```
public function registerSearch() {
    $this->getContainer()->query('OCP\ISearch')->registerProvider('OCA\YourApp\YourSearchClass', ['apps' => ['task']]);
}
```

Then in app.php you can do something like this:
$app = new Application();
$app->registerSearch();

PS: Every time you type OC::$server think again, you're probably making a mistake ;)

**Comment by raimund-schluessler** [comment]
Thanks for the extensive write-up :) I will try to implement your recommendations this weekend. Hopefully, I understood everything correctly ;)

**Comment by raimund-schluessler** [comment]
I have fixed the searchProvider but I don't know if everything now is as it should be. But at least the search is working again.

@Henni I will try to split up all the backend changes without things needed for subtasks and create a separate PR for this. The subtasks branch will be on hold until the new PR is merged so I can rebase subtasks onto master then.

@Raydiation I tried different things to wire up the SearchProvider and now use the version which is more or less also used here: https://github.com/libasys/pinit/blob/master/search/provider.php and created a separate SearchService for this. I am not quite sure if this is the best (your solution) but it does work. Maybe you could have a look at the backend PR once it is created.

**Comment by raimund-schluessler** [comment]
I split up the backend changes unrelated to Subtasks and created a new PR: #226.

@Henni Maybe you could have a look at it.

**Comment by Henni** [comment]
@raimund-schluessler I'll take a look on thursday after my exams.

**Comment by raimund-schluessler** [comment]
@Henni Thanks, and good look with your exams :)

**Comment by raimund-schluessler** [comment]
> Well, the current implementation works... But on my development server it takes very long to load (3~4s), and it takes literally forever on my productive server with many calendar entries (I stopped the request after 2 min).
There got to be a faster way to get the child tasks then looping through all tasks again and again (current implementation of Helper::getChildTasks()).

Loading time increase is neglectable now.

**Comment by raimund-schluessler** [comment]
@Henni I revised the implementation of drag and drop. It isnt perfect yet, but already a huge step forward.  The user already sees where the task will be droped. However, although dragging a task to be a subtask works, the animation is not functioning yet. I will try to fix this soon.

**Comment by Henni** [comment]
@raimund-schluessler i don't have time to test this at the moment. But looks like we can merge this very soon.
Will you be at the conference?

**Comment by raimund-schluessler** [comment]
@Henni Yes, I plan to be there on Saturday and Sunday.

**Comment by Henni** [comment]
@raimund-schluessler great! looking forward to meeting you there.

**Comment by raimund-schluessler** [comment]
Drag and Drop should work nicely in the views for a single calendar now. For the collections ('Important', 'Week', etc.) I still have to implement it.

**Comment by raimund-schluessler** [comment]
@Henni @jancborchardt This feature reached a state where I don't know any remaining issues. However, this doesn't guarantee that there are none :wink: So it would be nice if you could have a look at it and test it. We can discuss any issue you find at the conference. I would like to release a new version of the app after that.

**Comment by Henni** [comment]
Good to go :+1: 
Awesome work!

File: service/mapperhelper.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		$this->tasksMapper = $tasksMapper;
+		$this->helper = $helper;
+		$this->taskParser = $taskParser;
+	}
+
+	/**
+	 * creates a map of the tasks to find tasks by UID and RELATED-TO
+	 * checks the calendar entries and returns and array with two arrays as content:
+	 * [0] -> an array of the parsed tasks with the UIDs as keys
+	 * [1] -> an array of the child UIDs with the parent UIDs as keys
```
Comment (Henni) [inline service/mapperhelper.php]: TODO: fix comment

**Comment by sahne123** [comment]
drag n drop subtask to other lists is now working as expected.  :+1: 

Everything else is fine as well. great work!

**Comment by Henni** [comment]
Then it's ready to merge after the commits are squashed.

**Comment by raimund-schluessler** [comment]
Done, could be merged tomorrow.

**Pull request merged**